### PR TITLE
NTP server authentication. Issue #8794

### PR DESCRIPTION
**Pull request by vktg**
- [X] Redmine Issue: https://redmine.pfsense.org/issues/8794
- [X] Ready for review

Server side NTPv3 (RFC-1305) authentication

### CODE REVIEW

File: src/etc/inc/system.inc [hunk 1]
```diff
@@ -1780,10 +1786,28 @@ function system_ntp_configure() {
...
 	@unlink("{$g['varrun_path']}/ntpd.pid");
 
+	/* set NTP server authentication key */
+	if ($config['ntpd']['serverauth'] == 'yes') {
+		$ntpkeyscfg = "1 " . strtoupper($config['ntpd']['serverauthalgo']) . " " . base64_decode($config['ntpd']['serverauthkey']) . "\n";
+		if (!@file_put_contents("{$g['varetc_path']}/ntp.keys", $ntpkeyscfg)) {
+			log_error(sprintf(gettext("Could not open %s/ntp.keys for writing"), $g['varetc_path']));
+			return;
+		}
+	}
```
Comment (jim-p) [inline src/etc/inc/system.inc:1796]: If this auth is disabled, the code should check to see if the `ntp.keys` file is present and remove it.

That or unconditionally do an `unlink_if_exists()` on it before the `if` statement.

File: src/usr/local/www/services_ntpd.php [hunk 2]
```diff
@@ -487,6 +508,37 @@ function build_interface_list() {
 	'file'
 ))->addClass('btn-default');
 
+$section->addInput(new Form_Checkbox(
+	'serverauth',
+	'Enable NTP Server Authentication',
+	'Enable NTPv3 authentication (RFC 1305)',
+	$pconfig['serverauth']
+))->setHelp('Authentication allows the NTP client to verify that the server is in fact known ' .
+	    'and trusted and not an intruder intending accidentally or on purpose to masquerade as that server.');
```
Comment (jim-p) [inline src/usr/local/www/services_ntpd.php:517]: This seems a bit wordy/confusing. Could be simplified to:

> Authentication allows the NTP client to confirm it is communicating with the intended server, which protects against man-in-the-middle attacks.

**Review by jim-p — CHANGES_REQUESTED** [review]

File: src/etc/inc/system.inc (hunk 1 above)
Comment (vktg) [reply src/etc/inc/system.inc:1798]: Updated

File: src/usr/local/www/services_ntpd.php (hunk 2 above)
Comment (vktg) [reply src/usr/local/www/services_ntpd.php]: Done

**Comment by rbgarga** [comment]
@vktg please rebase your fork and resolve conflicts

**Review by jim-p — APPROVED** [review]

**Comment by vktg** [comment]
Rebased
